Framework/API/src/Run.cpp

```diff
@@ -344,12 +375,84 @@ void Run::setGoniometer(const Geometry::Goniometer &goniometer,
  * previously set Goniometer object as well as the angles
  * loaded in the run (if any).
  *
- * As of now, it uses the MEAN angle.
- *
  * @return 3x3 double rotation matrix
  */
 const Mantid::Kernel::DblMatrix &Run::getGoniometerMatrix() const {
-  return m_goniometer->getR();
+  return getGoniometerMatrix(0);
+}
+
+//-----------------------------------------------------------------------------------------------
+/// @return the number of goniometers's in this Run
+size_t Run::getNumGoniometers() const { return m_goniometers.size(); }
+
+//-----------------------------------------------------------------------------------------------
+/** Add a new Goniometer to this Run
+ *
+ * @param goniometer :: goniometer to add
+ * @return the index at which it was added
+ */
+size_t Run::addGoniometer(const Geometry::Goniometer &goniometer) {
+  m_goniometers.emplace_back(
+      std::make_unique<Geometry::Goniometer>(goniometer));
+  return m_goniometers.size() - 1;
+}
+
+//-----------------------------------------------------------------------------------------------
+/// Remove all goniometers on the Run
+void Run::clearGoniometers() { m_goniometers.clear(); }
+
+//-----------------------------------------------------------------------------------------------
+/** Get the Goniometer for the given run index
+ *
+ * @param index :: index of the run to get.
+ * @return goniometer
+ */
+const Geometry::Goniometer &Run::getGoniometer(const size_t index) const {
+  if (index >= m_goniometers.size())
+    throw std::invalid_argument(
```

Comment:
This would be better as a `std::out_of_range` like [`std::vector::at`](https://en.cppreference.com/w/cpp/container/vector/at)